infrastructure/src/main/java/org/corfudb/infrastructure/management/failuredetector/ClusterGraph.java

```diff
@@ -174,27 +177,57 @@ private boolean isLocalNode(NodeConnectivity node) {
      * - ClusterGraph is empty, which is an invalid state.
      * - the decision maker doesn't have connections, which is also impossible.
      *
+     * @param unhealthyNodes the list of unhealthy nodes from other detectors/agents (like read-only file system)
      * @return a decision maker node
      */
-    public Optional<NodeRank> getDecisionMaker() {
+    public Optional<NodeRank> getDecisionMaker(Set<String> healthyNodes) {
         log.trace("Get decision maker");
 
         NavigableSet<NodeRank> nodes = getNodeRanks();
 
+        List<NodeRank> unhealthyDecisionMakers = new ArrayList<>();
+        for (NodeRank node : nodes) {
+            if (!healthyNodes.contains(node.getEndpoint())) {
+                unhealthyDecisionMakers.add(node);
+            }
+        }
+
+        for (NodeRank unhealthyDecisionMaker : unhealthyDecisionMakers) {
+            nodes.remove(unhealthyDecisionMaker);
+        }
+
         if (nodes.isEmpty()) {
             log.error("Empty graph. Can't provide decision maker");
             return Optional.empty();
         }
```

Comment:
I think this can be simplified a bit:
```
        NavigableSet<NodeRank> healthyDecisionMakers = nodes.stream()
                .filter(healthyNodes::contains)
                .collect(Collectors.toCollection(TreeSet::new));
```
Then it would suffice to `pollFirst` from this set.

Reply:
done